File: apps/reading/src/api/ndShadowTwinPresentationCompetitionDrSourceAttachCompose.ts
/**
 * NotDiamond shadow advisory re-affirmation over recursive twin presentation
 * + competition DR source-attach Antiek-bench pack (pure).
 *
 * Operator vision: investigate NotDiamond as router — platform §16 REJECT as
 * production router; useful only as shadow/advisory beside operator model
 * decision. Re-affirms REJECT while twin presentation is ready.
 *
 * live_router_authorized always false.
 * twin_written / merge_executed / purchase_executed always false.
 * production_router_verdict always REJECT.
 */

import {
  composeNotDiamondShadowAdvisory,
  type NotDiamondShadowAdvisoryCompose,
  type NotDiamondShadowAdvisoryInput,
} from "./notDiamondShadowAdvisoryCompose";
import {
  composeRecursiveTwinPresentationCompetitionDrSourceAttach,
  type RecursiveTwinPresentationCompetitionDrSourceAttachCompose,
  type RecursiveTwinPresentationCompetitionDrSourceAttachInput,
} from "./recursiveTwinPresentationCompetitionDrSourceAttachCompose";

export interface NdShadowTwinPresentationCompetitionDrSourceAttachInput {
  nd_shadow: NotDiamondShadowAdvisoryInput;
  twin_presentation: Omit<
    RecursiveTwinPresentationCompetitionDrSourceAttachInput,
    "operator_ack"
  >;
  operator_ack: boolean;
  /**
   * When true (default), require nd shadow visible-or-kill-switch-honest AND
   * twin_presentation.pack_ready.
   */
  require_both?: boolean;
}

export interface NdShadowTwinPresentationCompetitionDrSourceAttachCompose {
  parent_asset_id: string;
  session_id: string;
  title: string;
  account_id: string;
  week_id: string;
  asset_id: string;
  nd_shadow: NotDiamondShadowAdvisoryCompose;
  twin_presentation: RecursiveTwinPresentationCompetitionDrSourceAttachCompose;
  pack_ready: boolean;
  live_router_authorized: false;
  twin_written: false;
  prompts_injected: false;
  merge_executed: false;
  live_dispatch_authorized: false;
  remote_fetched: false;
  backlog_mutated: false;
  purchase_executed: false;
  hosted: false;
  pdf_view_authorized: false;
  pdf_primary: false;
  secrets_stored: false;
  live_meter_read: false;
  store_mutated: false;
  suite_rewritten: false;
  live_execution_authorized: false;
  charge_executed: false;
  remote_index_queried: false;
  inventory_mutated: false;
  live_dispatched: false;
  pack_dispatched: false;
  draft_written: false;
  record_persisted: false;
  analysis_written: false;
  production_router_verdict: "REJECT";
  notes: string[];
  authority: "nd_shadow_twin_presentation_competition_dr_source_attach_compose_advisory";
}

function requireNonEmpty(value: unknown, name: string): string {
  if (typeof value !== "string" || !value.trim()) {
    throw new Error(`${name} must be a non-empty string`);
  }
  return value.trim();
}

/**
 * ND shadow REJECT re-affirmation on twin presentation + competition DR source-attach pack.
 * Never live-routes; never writes twin; never purchases.
 */
export function composeNdShadowTwinPresentationCompetitionDrSourceAttach(
  input: NdShadowTwinPresentationCompetitionDrSourceAttachInput,
): NdShadowTwinPresentationCompetitionDrSourceAttachCompose {
  if (!input || typeof input !== "object") {
    throw new Error("input must be an object");
  }
  if (typeof input.operator_ack !== "boolean") {
    throw new Error("operator_ack must be an explicit boolean");
  }
  if (!input.nd_shadow || typeof input.nd_shadow !== "object") {
    throw new Error("nd_shadow must be an object");
  }
  if (!input.twin_presentation || typeof input.twin_presentation !== "object") {
    throw new Error("twin_presentation must be an object");
  }

  const require_both =
    input.require_both === undefined ? true : input.require_both;
  if (typeof require_both !== "boolean") {
    throw new Error("require_both must be boolean when set");
  }

  const notes: string[] = [
    "production_router_verdict=REJECT — NotDiamond is not production router (§16)",
    "live_router_authorized=false · twin_written=false · merge_executed=false",
    "purchase_executed=false · remote_fetched=false",
  ];

  const nd_shadow = composeNotDiamondShadowAdvisory(input.nd_shadow);
  notes.push(...nd_shadow.notes.map((n) => `[nd_shadow] ${n}`));

  if (nd_shadow.production_router_verdict !== "REJECT") {
    throw new Error("invariant: production_router_verdict must be REJECT");
  }
  if (nd_shadow.live_router_authorized !== false) {
    throw new Error("invariant: live_router_authorized must be false");
  }

  const twin_presentation = composeRecursiveTwinPresentationCompetitionDrSourceAttach({
    ...input.twin_presentation,
    operator_ack: input.operator_ack,
  });
  notes.push(
    ...twin_presentation.notes.map((n) => `[twin_presentation] ${n}`),
  );

  const parent_asset_id = requireNonEmpty(
    twin_presentation.parent_asset_id,
    "parent_asset_id",
  );
  const session_id = requireNonEmpty(
    twin_presentation.session_id,
    "session_id",
  );
  const title = requireNonEmpty(twin_presentation.title, "title");
  const account_id = requireNonEmpty(
    twin_presentation.account_id,
    "account_id",
  );
  const week_id = requireNonEmpty(twin_presentation.week_id, "week_id");
  const asset_id = requireNonEmpty(twin_presentation.asset_id, "asset_id");

  const nd_gate =
    nd_shadow.production_router_verdict === "REJECT" &&
    nd_shadow.live_router_authorized === false;

  let pack_ready = false;
  if (require_both) {
    pack_ready =
      nd_gate &&
      twin_presentation.pack_ready === true &&
      twin_presentation.twin_written === false &&
      twin_presentation.merge_executed === false &&
      twin_presentation.purchase_executed === false &&
      twin_presentation.live_dispatch_authorized === false &&
      twin_presentation.remote_fetched === false &&
      twin_presentation.hosted === false &&
      twin_presentation.pdf_primary === false &&
      twin_presentation.production_router_verdict === "REJECT" &&
      input.operator_ack === true;
  } else {
    pack_ready =
      nd_gate &&
      input.operator_ack === true &&
      twin_presentation.purchase_executed === false &&
      twin_presentation.production_router_verdict === "REJECT" &&
      (twin_presentation.pack_ready === true ||
        nd_shadow.shadow_visible === true);
  }

  if (pack_ready) {
    notes.push(
      "pack_ready=true — ND shadow REJECT re-affirmed on twin presentation source-attach pack; still pure",
    );
  } else {
    notes.push(
      "pack_ready=false — nd_shadow, twin_presentation, or operator_ack gate open",
    );
  }

  if (
    nd_shadow.live_router_authorized !== false ||
    nd_shadow.production_router_verdict !== "REJECT" ||
    twin_presentation.twin_written !== false ||
    twin_presentation.merge_executed !== false ||
    twin_presentation.purchase_executed !== false ||
    twin_presentation.live_dispatch_authorized !== false ||
    twin_presentation.production_router_verdict !== "REJECT"
  ) {
    throw new Error("invariant: honesty flags must remain false / REJECT");
  }

  notes.push("live_router_authorized=false");
  notes.push("twin_written=false");
  notes.push("prompts_injected=false");
  notes.push("merge_executed=false");
  notes.push("live_dispatch_authorized=false");
  notes.push("remote_fetched=false");
  notes.push("backlog_mutated=false");
  notes.push("purchase_executed=false");
  notes.push("hosted=false");
  notes.push("pdf_view_authorized=false");
  notes.push("pdf_primary=false");
  notes.push("secrets_stored=false");
  notes.push("live_meter_read=false");
  notes.push("store_mutated=false");
  notes.push("suite_rewritten=false");
  notes.push("live_execution_authorized=false");
  notes.push("charge_executed=false");
  notes.push("remote_index_queried=false");
  notes.push("inventory_mutated=false");
  notes.push("live_dispatched=false");
  notes.push("pack_dispatched=false");
  notes.push("draft_written=false");
  notes.push("record_persisted=false");
  notes.push("analysis_written=false");
  notes.push("production_router_verdict=REJECT");

  return {
    parent_asset_id,
    session_id,
    title,
    account_id,
    week_id,
    asset_id,
    nd_shadow,
    twin_presentation,
    pack_ready,
    live_router_authorized: false,
    twin_written: false,
    prompts_injected: false,
    merge_executed: false,
    live_dispatch_authorized: false,
    remote_fetched: false,
    backlog_mutated: false,
    purchase_executed: false,
    hosted: false,
    pdf_view_authorized: false,
    pdf_primary: false,
    secrets_stored: false,
    live_meter_read: false,
    store_mutated: false,
    suite_rewritten: false,
    live_execution_authorized: false,
    charge_executed: false,
    remote_index_queried: false,
    inventory_mutated: false,
    live_dispatched: false,
    pack_dispatched: false,
    draft_written: false,
    record_persisted: false,
    analysis_written: false,
    production_router_verdict: "REJECT",
    notes,
    authority: "nd_shadow_twin_presentation_competition_dr_source_attach_compose_advisory",
  };
}

export function formatNdShadowTwinPresentationCompetitionDrSourceAttachSummary(
  c: NdShadowTwinPresentationCompetitionDrSourceAttachCompose,
): string {
  return (
    `pack_ready=${c.pack_ready} · ` +
    `nd_visible=${c.nd_shadow.shadow_visible} · ` +
    `suggested=${c.nd_shadow.suggested_model_id ?? "null"} · ` +
    `twin_presentation_ready=${c.twin_presentation.pack_ready} · ` +
    `view_mode=${c.twin_presentation.presentation.view_mode} · ` +
    `verdict=${c.production_router_verdict} · ` +
    `live_router_authorized=false · twin_written=false · purchase_executed=false`
  );
}

File: substrate/nd_shadow_twin_presentation_competition_dr_source_attach_compose.py
"""NotDiamond shadow REJECT re-affirmation over twin presentation pack.

live_router_authorized always False.
twin_written / merge_executed / purchase_executed always False.
production_router_verdict always REJECT.
"""

from __future__ import annotations

from dataclasses import dataclass
from typing import Any

from substrate.notdiamond_shadow_advisory_compose import (
    NotDiamondShadowAdvisoryCompose,
    NotDiamondShadowAdvisoryComposeError,
    compose_notdiamond_shadow_advisory,
)
from substrate.recursive_twin_presentation_competition_dr_source_attach_compose import (
    RecursiveTwinPresentationCompetitionDrSourceAttachCompose,
    RecursiveTwinPresentationCompetitionDrSourceAttachComposeError,
    compose_recursive_twin_presentation_competition_dr_source_attach,
)


class NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(ValueError):
    """Fail-closed validation for ND shadow + twin presentation pack."""


@dataclass(frozen=True)
class NdShadowTwinPresentationCompetitionDrSourceAttachCompose:
    parent_asset_id: str
    session_id: str
    title: str
    account_id: str
    week_id: str
    asset_id: str
    nd_shadow: NotDiamondShadowAdvisoryCompose
    twin_presentation: RecursiveTwinPresentationCompetitionDrSourceAttachCompose
    pack_ready: bool
    live_router_authorized: bool
    twin_written: bool
    prompts_injected: bool
    merge_executed: bool
    live_dispatch_authorized: bool
    remote_fetched: bool
    backlog_mutated: bool
    purchase_executed: bool
    hosted: bool
    pdf_view_authorized: bool
    pdf_primary: bool
    secrets_stored: bool
    live_meter_read: bool
    store_mutated: bool
    suite_rewritten: bool
    live_execution_authorized: bool
    charge_executed: bool
    remote_index_queried: bool
    inventory_mutated: bool
    live_dispatched: bool
    pack_dispatched: bool
    draft_written: bool
    record_persisted: bool
    analysis_written: bool
    production_router_verdict: str
    notes: tuple[str, ...]
    authority: str

    def to_dict(self) -> dict[str, Any]:
        return {
            "parent_asset_id": self.parent_asset_id,
            "session_id": self.session_id,
            "title": self.title,
            "account_id": self.account_id,
            "week_id": self.week_id,
            "asset_id": self.asset_id,
            "nd_shadow": self.nd_shadow.to_dict(),
            "twin_presentation": self.twin_presentation.to_dict(),
            "pack_ready": self.pack_ready,
            "live_router_authorized": False,
            "twin_written": False,
            "prompts_injected": False,
            "merge_executed": False,
            "live_dispatch_authorized": False,
            "remote_fetched": False,
            "backlog_mutated": False,
            "purchase_executed": False,
            "hosted": False,
            "pdf_view_authorized": False,
            "pdf_primary": False,
            "secrets_stored": False,
            "live_meter_read": False,
            "store_mutated": False,
            "suite_rewritten": False,
            "live_execution_authorized": False,
            "charge_executed": False,
            "remote_index_queried": False,
            "inventory_mutated": False,
            "live_dispatched": False,
            "pack_dispatched": False,
            "draft_written": False,
            "record_persisted": False,
            "analysis_written": False,
            "production_router_verdict": "REJECT",
            "notes": list(self.notes),
            "authority": (
                "nd_shadow_twin_presentation_competition_dr_source_attach_compose_advisory"
            ),
        }


def _require_nonempty(value: object, *, field: str) -> str:
    if not isinstance(value, str) or not value.strip():
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(
            f"{field} must be a non-empty string"
        )
    return value.strip()


def compose_nd_shadow_twin_presentation_competition_dr_source_attach(
    *,
    nd_shadow: object,
    twin_presentation: object,
    operator_ack: object,
    require_both: object | None = None,
) -> NdShadowTwinPresentationCompetitionDrSourceAttachCompose:
    """ND shadow REJECT on twin presentation pack. Never live-routes."""
    if not isinstance(operator_ack, bool):
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(
            "operator_ack must be an explicit boolean"
        )
    if not isinstance(nd_shadow, dict):
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(
            "nd_shadow must be an object"
        )
    if not isinstance(twin_presentation, dict):
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(
            "twin_presentation must be an object"
        )

    require = True if require_both is None else require_both
    if not isinstance(require, bool):
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(
            "require_both must be boolean when set"
        )

    notes: list[str] = [
        "production_router_verdict=REJECT — NotDiamond is not production "
        "router (§16)",
        "live_router_authorized=false · twin_written=false · "
        "merge_executed=false",
        "purchase_executed=false · remote_fetched=false",
    ]

    try:
        nd = compose_notdiamond_shadow_advisory(
            selected_model_id=nd_shadow.get("selected_model_id"),
            nd_recommended_model_id=nd_shadow.get("nd_recommended_model_id"),
            kill_switch_on=nd_shadow.get("kill_switch_on"),
            confidence=nd_shadow.get("confidence"),
            task=nd_shadow.get("task"),
            inventory_model_ids=nd_shadow.get("inventory_model_ids"),
        )
    except NotDiamondShadowAdvisoryComposeError as e:
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(str(e)) from e
    notes.extend(f"[nd_shadow] {n}" for n in nd.notes)

    if nd.production_router_verdict != "REJECT":
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(
            "invariant: production_router_verdict must be REJECT"
        )
    if nd.live_router_authorized is not False:
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(
            "invariant: live_router_authorized must be false"
        )

    try:
        twin = compose_recursive_twin_presentation_competition_dr_source_attach(
            twin=twin_presentation.get("twin"),
            presentation=twin_presentation.get("presentation"),
            competition_pack=twin_presentation.get("competition_pack"),
            operator_ack=operator_ack,
            require_both=twin_presentation.get("require_both"),
        )
    except RecursiveTwinPresentationCompetitionDrSourceAttachComposeError as e:
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(str(e)) from e
    notes.extend(f"[twin_presentation] {n}" for n in twin.notes)

    parent = _require_nonempty(twin.parent_asset_id, field="parent_asset_id")
    session = _require_nonempty(twin.session_id, field="session_id")
    title = _require_nonempty(twin.title, field="title")
    account = _require_nonempty(twin.account_id, field="account_id")
    week = _require_nonempty(twin.week_id, field="week_id")
    asset = _require_nonempty(twin.asset_id, field="asset_id")

    nd_gate = (
        nd.production_router_verdict == "REJECT"
        and nd.live_router_authorized is False
    )

    if require:
        pack_ready = (
            nd_gate
            and twin.pack_ready is True
            and twin.twin_written is False
            and twin.merge_executed is False
            and twin.purchase_executed is False
            and twin.live_dispatch_authorized is False
            and twin.production_router_verdict == "REJECT"
            and operator_ack is True
        )
    else:
        pack_ready = (
            nd_gate
            and operator_ack is True
            and twin.purchase_executed is False
            and twin.production_router_verdict == "REJECT"
            and (twin.pack_ready is True or nd.shadow_visible is True)
        )

    if pack_ready:
        notes.append(
            "pack_ready=true — ND shadow REJECT re-affirmed on twin "
            "presentation pack; still pure"
        )
    else:
        notes.append(
            "pack_ready=false — nd_shadow, twin_presentation, or operator_ack "
            "gate open"
        )

    if (
        nd.live_router_authorized is not False
        or nd.production_router_verdict != "REJECT"
        or twin.twin_written is not False
        or twin.merge_executed is not False
        or twin.purchase_executed is not False
        or twin.live_dispatch_authorized is not False
        or twin.production_router_verdict != "REJECT"
    ):
        raise NdShadowTwinPresentationCompetitionDrSourceAttachComposeError(
            "invariant: honesty flags must remain false / REJECT"
        )

    notes.extend(
        (
            "live_router_authorized=false",
            "twin_written=false",
            "prompts_injected=false",
            "merge_executed=false",
            "live_dispatch_authorized=false",
            "remote_fetched=false",
            "backlog_mutated=false",
            "purchase_executed=false",
            "hosted=false",
            "pdf_view_authorized=false",
            "pdf_primary=false",
            "secrets_stored=false",
            "live_meter_read=false",
            "store_mutated=false",
            "suite_rewritten=false",
            "live_execution_authorized=false",
            "charge_executed=false",
            "remote_index_queried=false",
            "inventory_mutated=false",
            "live_dispatched=false",
            "pack_dispatched=false",
            "draft_written=false",
            "record_persisted=false",
            "analysis_written=false",
            "production_router_verdict=REJECT",
        )
    )

    return NdShadowTwinPresentationCompetitionDrSourceAttachCompose(
        parent_asset_id=parent,
        session_id=session,
        title=title,
        account_id=account,
        week_id=week,
        asset_id=asset,
        nd_shadow=nd,
        twin_presentation=twin,
        pack_ready=pack_ready,
        live_router_authorized=False,
        twin_written=False,
        prompts_injected=False,
        merge_executed=False,
        live_dispatch_authorized=False,
        remote_fetched=False,
        backlog_mutated=False,
        purchase_executed=False,
        hosted=False,
        pdf_view_authorized=False,
        pdf_primary=False,
        secrets_stored=False,
        live_meter_read=False,
        store_mutated=False,
        suite_rewritten=False,
        live_execution_authorized=False,
        charge_executed=False,
        remote_index_queried=False,
        inventory_mutated=False,
        live_dispatched=False,
        pack_dispatched=False,
        draft_written=False,
        record_persisted=False,
        analysis_written=False,
        production_router_verdict="REJECT",
        notes=tuple(notes),
        authority="nd_shadow_twin_presentation_competition_dr_source_attach_compose_advisory",
    )


def format_nd_shadow_twin_presentation_competition_dr_source_attach_summary(
    c: NdShadowTwinPresentationCompetitionDrSourceAttachCompose,
) -> str:
    return (
        f"pack_ready={c.pack_ready} · "
        f"nd_visible={c.nd_shadow.shadow_visible} · "
        f"suggested={c.nd_shadow.suggested_model_id} · "
        f"twin_presentation_ready={c.twin_presentation.pack_ready} · "
        f"view_mode={c.twin_presentation.presentation.view_mode} · "
        f"verdict={c.production_router_verdict} · "
        f"live_router_authorized=false · twin_written=false · "
        f"purchase_executed=false"
    )


__all__ = [
    "NdShadowTwinPresentationCompetitionDrSourceAttachCompose",
    "NdShadowTwinPresentationCompetitionDrSourceAttachComposeError",
    "compose_nd_shadow_twin_presentation_competition_dr_source_attach",
    "format_nd_shadow_twin_presentation_competition_dr_source_attach_summary",
]
